File: StopWatch_aac/app/src/main/java/com/thomaskuenneth/stopwatch_aac/StopwatchLifecycleObserver.java
package com.thomaskuenneth.stopwatch_aac;

import android.arch.lifecycle.Lifecycle;
import android.arch.lifecycle.LifecycleObserver;
import android.arch.lifecycle.OnLifecycleEvent;

import java.util.Timer;
import java.util.TimerTask;

public class StopwatchLifecycleObserver implements LifecycleObserver {

    private final StopwatchViewModel model;

    private Timer timer;
    private TimerTask timerTask;

    StopwatchLifecycleObserver(StopwatchViewModel model) {
        this.model = model;
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_RESUME)
    public void startTimer() {
        timer = new Timer();
        Boolean running = model.running.getValue();
        if ((running != null) && (running)) {
            scheduleAtFixedRate();
        }
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_PAUSE)
    public void stopTimer() {
        timer.cancel();
    }

    void stop() {
        timerTask.cancel();
    }

    void scheduleAtFixedRate() {
        Long diff = model.diff.getValue();
        if (diff != null) {
            model.started.setValue(System.currentTimeMillis() - diff);
        }
        timerTask = new TimerTask() {
            @Override
            public void run() {
                Long started = model.started.getValue();
                if (started != null) {
                    model.diff.postValue(System.currentTimeMillis() - started);
//                    handler.post(()
//                            -> model.diff.setValue(System.currentTimeMillis() - started));
                }
            }
        };
        timer.scheduleAtFixedRate(timerTask, 0, 200);
    }
}

File: xamarin/Stopwatch_AAC/Stopwatch_AAC/StopwatchLifecycleObserver.cs
﻿using Android.Arch.Lifecycle;
using Android.OS;
using Android.Util;
using Java.Interop;
using Java.Lang;
using Java.Util;

namespace Stopwatch_AAC
{
    public class StopwatchLifecycleObserver : Object, ILifecycleObserver
    {
        const string TAG = "StopwatchLifecycleObserver";

        readonly StopwatchViewModel model;
        readonly Handler handler;

        Timer timer;
        TimerTask timerTask;

        public StopwatchLifecycleObserver(StopwatchViewModel model,
                                          Handler handler)
        {
            this.model = model;
            this.handler = handler;
        }

        [Lifecycle.Event.OnAny]
        [Export]
        public void Hello(ILifecycleOwner owner,
                          Lifecycle.Event evt)
        {
            Log.Debug(TAG, owner.ToString());
            Log.Debug(TAG, evt.ToString());
        }

        [Lifecycle.Event.OnResume]
        [Export]
        public void StartTimer()
        {
            timer = new Timer();
            if (model.Running)
            {
                ScheduleAtFixedRate();
            }
        }

        [Lifecycle.Event.OnPause]
        [Export]
        public void StopTimer()
        {
            timer.Cancel();
        }

        public void Stop()
        {
            timerTask.Cancel();
        }

        public void ScheduleAtFixedRate()
        {
            model.Started = JavaSystem.CurrentTimeMillis() - model.Diff;
            timerTask = new MyTimerTask(handler, model);
            timer.ScheduleAtFixedRate(timerTask, 0, 200);
        }
    }
}
